### ISSUE TITLE
General Development Talk

### ISSUE THREAD

**Issue by ravenwilde** [kisanjong/rotalabutterfly#7]
This thread is reserved for discussion between developers regarding code hand-offs etc. 

Any bugs, new features, or calculation errors should be put into individual tickets.

**Comment by ravenwilde** [comment]
@fablau - carrying this discussion over from the old thread:

How do we want to handle messages - do you want to compile them with the appropriate embedded variables server side? Or would you like me to try and handle it in the javascript?

**Comment by ravenwilde** [comment]
@fablau PS. I've set the 'Round to...' select dropdown to default to 2.

**Comment by fablau** [comment]
Great to be on this new thread!

> How do we want to handle messages - do you want to compile them with the appropriate embedded variables server side? Or would you like me to try and handle it in the javascript?

Either way, as you prefer. For me is pretty simple, but if you have already something in place, and you like to do it in JS, please, go ahead. Let me know, it's your call! :)

**Comment by kisanjong** [comment]
Ui ver 1.1 now on test server and looking great :)

**Comment by ravenwilde** [comment]
@fablau - I had an 'aha!' moment last night and realized that I could handle the output message variable interpolation with a string replace function. So I'll take care of this on my end.

**Comment by ravenwilde** [comment]
I've opened a 'Calculation Testing' Issue that you (Jason) can link to on the forums if you want to enlist people as testers. I figure its probably easier (for them) if we have them report any problems they find in that one thread rather than have them open new issues.

Also, please chime in on that thread if I forgot to list something else that they should report.

**Comment by fablau** [comment]
> PS. I've set the 'Round to...' select dropdown to default to 2.

I forgot this... great, thank you Jen!

**Comment by fablau** [comment]
> I had an 'aha!' moment last night and realized that I could handle the output message variable interpolation with a string replace function. So I'll take care of this on my end.

Fantastic!

**Comment by fablau** [comment]
> Ui ver 1.1 now on test server and looking great :)

Yes, working great indeed. Great job Jennifer!

**Comment by kisanjong** [comment]
Links posted on the Barr report and I'll speak to a few people asap for beta testers.

**Comment by kisanjong** [comment]
Should we open an issue for the high charts feature that was used in wets calculator? 

http://www.highcharts.com/demo

**Comment by ravenwilde** [comment]
Yes. I will do that. I'm still trying to figure out how he was generating that range chart - I'm not entirely sure that highcharts was used for that.

**Comment by kisanjong** [comment]
I'm not positive it was high charts but its mentioned in his readme file at Flores.

**Comment by ravenwilde** [comment]
I think he was using highcharts for the longterm modeling but the range chart may have been homegrown.... I've been looking through the code, but its ruby, so its taking a while ;)

**Comment by fablau** [comment]
> I think he was using highcharts for the longterm modeling but the range chart may have been homegrown.... I've been looking through the code, but its ruby, so its taking a while ;)

Jennifer, I think the code controlling the chart is at the bottom of the original nutes.rb (from line 301, "fancy graphs"). I left that part of the code out from my translation to PHP.

**Comment by fablau** [comment]
> Links posted on the Barr report and I'll speak to a few people asap for beta testers.

Great, thank you Jason!

**Comment by ravenwilde** [comment]
> Jennifer, I think the code controlling the chart at the bottom of the original nutes.rb (from line 301). I left that code out from my translation to PHP.

Yeah, that's what I've been looking at - I think you may have to generate some of the vars and pass them off to me.

**Comment by fablau** [comment]
Yes, I think I can do that pretty easily. Looks like he was handling template place-holders directly there. I could translate those into simple vars, and then give them back to you in the AJAX response... thoughts?

**Comment by fablau** [comment]
The only issue could be with loops... we could end up having a lot of variables, and I don't think that would be the right way to do it. That should be done via JS, maybe having the starting and ending points only? I'll think a little bit more about this.

**Comment by ravenwilde** [comment]
Let me look at the code more in depth and I'll have a more cogent response for you later tonight.

**Comment by fablau** [comment]
> Let me look at the code more in depth and I'll have a more cogent response for you later tonight.

OK!

**Comment by kisanjong** [comment]
Some one mentioned multiple selection for example the whole ei..

They can select trace +n +p +k and get results as a Printable list for all?

**Comment by fablau** [comment]
> Some one mentioned multiple selection for example the whole ei..
> They can select trace +n +p +k and get results as a Printable list for all?

That could be done easily UI input side, but the whole work would be JS side, therefore Jennifer's... she should re-iterate more calls to the PHP backend for each requested compound, then build the output list. Maybe keep this as a future feature?

**Comment by ravenwilde** [comment]
> Maybe keep this as a future feature?

Yeah... this is something we could do later on. For now I'd like to focus on replicating the features of wet's calculator. Once we get all of those taken care of then we can look at building out something like this.

**Comment by kisanjong** [comment]
I totally agree Jennifer that new features will come after.

**Comment by kisanjong** [comment]
Hello both :) how things going? :)

**Comment by kisanjong** [comment]
Looking good Jennifer :) @fablau will need to confirm the language is correct but from what I can see its working fine on the ui just the results are still English.

**Comment by fablau** [comment]
> Looking good Jennifer :) @fablau will need to confirm the language is correct but from what I can see its working fine on the ui just the results are still English.

Jason, do you want me to look at the Italian template file? Please, let me know. Thanks.

**Comment by ravenwilde** [comment]
Hi guys - I've in the middle of transitioning to a new job (Yaaaay!) so I had to put this on hold for a bit. I'm also moving all my work over to a different computer so I need to set my dev environment up again. I've got it mostly there but I need to do that whole, PEAR, PECL, YAML thing again. I'll probably be back on track and working on this again by early next week.

@fablau - If you want to start work on the Italian language file you should look at the one that I pushed along with the UI branch - its located here: https://github.com/kisanjong/rotalabutterfly/blob/ui/json/it.json
I had to change it from the original a bit to work better with our current UI... and I used Google Translate, so I am 100% certain that it needs improving. :)

**Comment by kisanjong** [comment]
Yes please Fab as Jennifer said :)

**Comment by kisanjong** [comment]
That's great news Jennifer congratulations on the new job :)

**Comment by kisanjong** [comment]
Guys I have a couple of vb4 licences that are not being used so do you think it's worth me setting one up on rotalabutterfly so we can use it for feedback and development?

Keeping all the feedback and translations etc from testers and interested users in one forum rather than all over the place.

It would also be handy for feature ideas, new commercial product information etc.

?

Thanks

**Comment by kisanjong** [comment]
you both have an email :)

**Comment by kisanjong** [comment]
I've moved all files from the /test/ to the root and linked wets original one in the header menu along with the forum link.

now hosted at rotalabutterfly.com

Jennifer i have not added these links to github as i don't want to interfere with your current work but would like them adding when possible please.

Thanks :)

**Comment by fablau** [comment]
> @fablau - If you want to start work on the Italian language file you should look at the one that I pushed along with the UI branch - its located here: https://github.com/kisanjong/rotalabutterfly/blob/ui/json/it.json
> I had to change it from the original a bit to work better with our current UI... and I used Google Translate, so I am 100% certain that it needs improving. :)

Ok, I can work on that. Just to be sure, once I have modified the it.json inside the UI branch, can I just "Commit to UI"?

**Comment by fablau** [comment]
> I've in the middle of transitioning to a new job (Yaaaay!)

Congrats indeed!

**Comment by fablau** [comment]
> you both have an email :)

Ok, I'll check it out asap. Thanks Jason.

**Comment by fablau** [comment]
Jennifer, I am afraid you missed my question above: once I update the it.json file, can I just "commit to UI" inside GIT?

**Comment by ravenwilde** [comment]
Sorry, @fablau - Yes, just commit your changes to that file to the UI branch and I can pull them from there. Thanks for your patience - I should be back to work on this project tomorrow.

**Comment by fablau** [comment]
Thank you Jennifer, just updated the file and committed, please, let me know if that's ok.

I have found a lot of mistakes, wrong phrases, etc... a little mess. Hopefully now is more correct :)

Also, I have noticed a discrepancy with the corresponding English template, on line 42 the English template (en.json) includes the following line:

`"dose": "`Your addition of ${dose_amount} to your ${tank_size} aquarium adds`",`

Whereas in the Italian template was this way:

`"dose": "Aggiungendo %1 al tuo acquario di %2 ti d√†",`

So, I have changed it to:

`"dose": "Aggiungendo ${dose_amount} al tuo acquario di ${tank_size} ti d√†",`

But I guess the following lines need ot be changed as well, in both templates (probably all of them) to use the "newer"  ${dose_amount} instead than %1, etc. (I guess so... am I right Jennifer?)

Best,
Fab.

**Comment by kisanjong** [comment]
I think you are correct about the %1, your commit was to this file? 
https://github.com/kisanjong/rotalabutterfly/blob/ui/json/it.json

I'm asking as I don't see the commit and the last commit date is older than today.

Thanks

**Comment by fablau** [comment]
Yes Jason, that's the correct file now (I had to Sync again right now!)

**Comment by kisanjong** [comment]
Got it thanks :)

**Comment by kisanjong** [comment]
sorry i pressed the wrong button when posting and closed the issue by mistake..

the ui seems fine fab, but the results are still in English unless this is still in development?

Italian file now added to http://rotalabutterfly.com

**Comment by fablau** [comment]
> the ui seems fine fab, but the results are still in English unless this is still in development?

Jennifer can help on that?

**Comment by ravenwilde** [comment]
Still in development - I'm going to put in some work on this later today, but, just so you guys know, I'm going to be tied up from Friday thru Monday and probably won't have much to push until Wednesday of next week. Sorry for being slow - things are kind of chaotic for me right now with the job transition.

**Comment by fablau** [comment]
No worries Jen, this is just for fun! Take it easy :)

**Comment by kisanjong** [comment]
Hi Jen as Fab said :) the new job is far more important.

**Comment by kisanjong** [comment]
@ravenwilde I've added a few more issues for you to see, the KCI has been fixed and changed in the compounds file of the php branch.

**Comment by ravenwilde** [comment]
Hey @kisanjong and @fablau - I'm back to working on this now, just finishing setting up my dev environment on my new (to me) computer and catching up on what's been going on here discussion-wise. Might take me a few days before I have anything to push but I'm back in touch now.

**Comment by kisanjong** [comment]
Hey Jen welcome back :), I see you have read my updated issues, some bugs to fix now and some enhancements to add later..

Looking forward to the updates.

**Comment by kisanjong** [comment]
I also added / changed the compounds file data in the ui branch

**Comment by fablau** [comment]
Nice to see you here Jen. Please, keep us posted with the development. Thanks!

**Comment by kisanjong** [comment]
i've uploaded the edits to http:// rotalabutterfly.com for testing

**Comment by fablau** [comment]
Great, thank you Jason.

**Comment by ravenwilde** [comment]
Thanks for that Jason, I just pushed updates to dev and then merged all that in from master. @kisanjong when you have the time - please set master as the default branch since the site is technically live now.

Moving forward the general procedure should be:

**master branch:** is always up to date with whatever is live at rotalabutterfly.com and whenever you go to update the live site you should get the files from the master branch

**dev branch:** is reserved for code that is still being tested by us and is not considered ready to go 'live' 

**feature branches (like 'ui' or 'php'):** are for us to do our individual work, etc. These get merged into dev for testing.

And thanks for all your hard work guys! This is going great!

**Comment by ravenwilde** [comment]
@kisanjong I just saw this:

> I've moved all files from the /test/ to the root and linked wets original one in the header menu along with the forum link.
> 
> now hosted at rotalabutterfly.com
> 
> Jennifer i have not added these links to github as i don't want to interfere with your current work but would like them adding when possible please.

Feel free to check these changes into the dev branch and then merge them into master. Thanks!

**Comment by kisanjong** [comment]
Thanks Jen :)

Master is now default and I'll look at the other changes now :+1: 

Thank you for your hard work as well its looking great.

**Comment by ravenwilde** [comment]
You too! This wouldn't have happened with out you!

**Comment by kisanjong** [comment]
Thanks :) 

current ui is now available at http://rotalabutterfly.com/test/

**Comment by fablau** [comment]
Thank you for allowing me to be part of this project! Great work guys!

**Comment by kisanjong** [comment]
Hi Fab & Jen

After numerous requests can we start on the results of my dose part of the calculator?

Including the graphical display.

Thanks
Jason

**Comment by fablau** [comment]
Hi Jason, sorry for my late reply.

The back-end should already work for that... am I correct Jennifer?

**Comment by ravenwilde** [comment]
@fablau - as far as I can tell this isn't working - all the results that come back when I try to get the result of a dose are zero or null.

EDIT: I'm also not sure that Target is is working either.

**Comment by fablau** [comment]
Hi Jen. The "Dose to reach a target" seems to return some results... we'll need to check if they are correct though.

The "Result of my dose" doesn't return anything, you are right. I'll look into that asap.

@kisanjong - I can't find the URL to the old-working calculator, could you please remind me? I'd need to compare results if possible.

Thanks!

**Comment by kisanjong** [comment]
Hi Fab link is in the header of the new calculator ;)

WET'S OLD CALCULATOR
http://yanc.rotalabutterfly.com/

**Comment by fablau** [comment]
Ouch!! Sorry missed that!!

Thank you very much, I will look into everything ASAP.
Thanks!

Fabrizio

**Comment by kisanjong** [comment]
Hehe no problem :)

And thanks

**Comment by fablau** [comment]
Thanks Jason… found what the problem is, I’ll get back to Jennifer asap.

Thanks!

Fab.

**Comment by fablau** [comment]
Ok. Jennifer, looks like the problem are the variables passed to the back-end. They should be the same sent from the old system.

For example, if you select the following values:

Tank size: 10 US gallons
Method: DIY
Nutrient: K2HPO4
Using: Dry Dosing
I am Calculating For: Result of my Dose
I am adding: 50 ml (it should be mg though!)

And then hit calculate, here is what the new calculator UI is passing to the backend:

---

calc_for: result
compound: K2HPO4
dose_amount: 50
dose_units: milliliter
method: dry
round_to: 2
source: diy
tank_units: gal
tank_vol: 10

---

Whereas, the old calculator is passing the following values instead:

---

calc_for: dump <--- This is the correct value
compound: K2HPO4
dose_amount: 50
dose_units: mg <--- This is the correct value
method: dry
round_to: 2
source: diy
tank_units: gal
tank_vol: 10

---

Please, let me know if by changing the values as above makes the trick. Thanks.

**Comment by ravenwilde** [comment]
Ahhhh... okay, I'll fix that problem tonight or tomorrow and push a fix for you.

**Comment by fablau** [comment]
Sounds good! Thank you Jen :)

**Comment by ravenwilde** [comment]
Question though, on your end can you change 'dump' to 'result'? Dump just doesn't make sense semantically and that kind of bothers me :)

**Comment by fablau** [comment]
Just done Jen. Let me know if it works.

**Comment by kisanjong** [comment]
Hello Fab & Jen how's its going? 

@ravenwilde what's the next stage your planing?

Thanns

**Comment by fablau** [comment]
Hi Jason. I was waiting for Jennifer to tell me if the backend is now working correctly for the "Result of my dose". What else do I need to take care of? I lost some track here... thanks for reminding me!

**Comment by kisanjong** [comment]
Hi Fab

No worries been busy myself :)

 I think that's the next stage, once that's confirmed its onto the result of my dose then a few beta checks from guys on the forum and then onto the other features?

Thanks

**Comment by fablau** [comment]
I am ok with that Jason! Let's wait for Jennifer. Thanks!

**Comment by kisanjong** [comment]
Hello,

Burr_740 has just posted in the General Discussion forum of rotalabutterfly.com Forums under the title of Getting very different results between old and new calc....

This thread is located at http://rotalabutterfly.com/forum/showthread.php/12-Getting-very-different-results-between-old-and-new-calc

Here is the message that has just been posted:

---

Hi guys, thanks a lot for your work on this.

You may already be aware, but Im also getting a difference between the old and the new for csm-b.

 Calculating for 65 gal, EI, dry dose: Old says 1 gram, new says 1.88.

Just thought you may need to know. Thanks again

---

**Comment by fablau** [comment]
Hi Jason. Yes, I see the difference... that's weird... I'll look into that as soon as possible.

**Comment by fablau** [comment]
Ok, I need to look into this in depth, but I think the problem could be that the old calculator rounds results, whereas our new one is more precise (2 decimal points). I'll confirm this as soon as I can look at the code, I am currently out town.

**Comment by ravenwilde** [comment]
Hi guys, sorry, I've been busy with work and family stuff. I'm going to make a concerted effort to work on this this week - I will be in touch.

**Comment by kisanjong** [comment]
Thanks Both :)

**Comment by fablau** [comment]
Glad to see you here Jen! Jason: you are welcome.

I have looked into the rounding issue above, and I can confirm that our new calculator shows more precise results according to the decimal digits defined in the calculator form, whereas Wet's calculator ignored that in some way (!!!) and rounded everything to the integer despite shows 0 as decimal digit (i.e. 1.0 instead of 1.88, or 3.0 instead of 3.01, etc)

**Comment by fablau** [comment]
Jason, Jennifer, yesterday I used our new calculator for my own tank needs, and we would need to change the output when the user select "a solution" as a result compared to "dry dosing". The output may be confusing to the user, it tells to add a certain amount "to your tank" instead than to "a container" of the defined volume.

**Comment by kisanjong** [comment]
@ravenwilde how are things? where are you ;)

**Comment by ravenwilde** [comment]
@kisanjong I've been working on overhauling part of the interface so we can print out charts. I don't have anything to push yet because its all in pieces. I actually have time this weekend to devote to this project so I will get back to you then.

**Comment by kisanjong** [comment]
Hi Jen, that's great news, I'm looking forward to the charts, thanks
